### PR TITLE
SNOW-821644 add dataframe alias support

### PR DESCRIPTION
**Pull request by sfc-gh-zli**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   *Replace this with a short description of how your code change solves the related issue.*

## Pre-review checklist

(For Snowflake employees)

- [ ] This change has passed precommit
- [ ] I have reviewed code coverage report for my PR in  ([Sonarqube](https://sonarqube.int.snowflakecomputing.com/dashboard?id=snowpark))

### CODE REVIEW

File: src/main/scala/com/snowflake/snowpark/internal/analyzer/SnowflakePlanNode.scala [hunk 1]
```diff
@@ -18,6 +19,15 @@ private[snowpark] trait LogicalPlan {
     (analyzedPlan, analyzer.getAliasMap)
   }
 
+  var dfAliasMap: MMap[String, Seq[Attribute]] = MMap.empty
```
Comment (sfc-gh-bli) [inline src/main/scala/com/snowflake/snowpark/internal/analyzer/SnowflakePlanNode.scala:22]: `LogicalPlan` should be immutable. So don't change it after creation, move this to the constructor like I mentioned above

File: src/main/scala/com/snowflake/snowpark/DataFrame.scala [hunk 2]
```diff
@@ -2791,7 +2808,8 @@ class DataFrame private[snowpark] (
 
   // utils
   private[snowpark] def resolve(colName: String): NamedExpression = {
-    val normalizedColName = quoteName(colName)
+    val (aliasColName, aliasOutput) = resolveAlias(colName, output)
```
[the remark this answers is not in the document]
Comment (sfc-gh-zli) [reply src/main/scala/com/snowflake/snowpark/DataFrame.scala:2811]: We need to quote the column name and check for name conflicts, so it's better to return the column name and output list.

File: src/main/scala/com/snowflake/snowpark/internal/analyzer/SnowflakePlanNode.scala [hunk 3]
```diff
@@ -192,6 +206,17 @@ private[snowpark] case class Sort(order: Seq[SortOrder], child: LogicalPlan) ext
     Sort(order, _)
 }
 
+private[snowpark] case class DataframeAlias(alias: String, child: LogicalPlan)
+  extends UnaryNode {
+  dfAliasMap += (alias -> child.getSnowflakePlan.get.output)
```
Comment (sfc-gh-bli) [inline src/main/scala/com/snowflake/snowpark/internal/analyzer/SnowflakePlanNode.scala:211]:
```scala
override lazy val dfAliasMap = child.dfAliasMap + (alias -> child.getSnowflakePlan.get.output)
```

File: src/main/scala/com/snowflake/snowpark/internal/analyzer/binaryPlanNodes.scala [hunk 4]
```diff
@@ -14,8 +14,10 @@ private[snowpark] abstract class BinaryNode extends LogicalPlan {
   lazy protected val analyzedRight: LogicalPlan = right.analyzed
 
   lazy override protected val analyzer: ExpressionAnalyzer =
-    ExpressionAnalyzer(left.aliasMap, right.aliasMap)
+    ExpressionAnalyzer(left.aliasMap, right.aliasMap, dfAliasMap)
 
+  addToDataframeAliasMap(left)
+  addToDataframeAliasMap(right)
```
Comment (sfc-gh-bli) [inline src/main/scala/com/snowflake/snowpark/internal/analyzer/binaryPlanNodes.scala:20]: override lazy val dfAliasMap = left.dfAliasMap ++ right.dfAliasMap

File: src/test/scala/com/snowflake/snowpark_test/DataFrameAliasSuite.scala [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+  test("Test for alias conflict") {
+    createTable(tableName1, "id int, num int")
+    createTable(tableName2, "id int, num int")
+    val df1 = session.table(tableName1).alias("A")
+    val df2 = session.table(tableName2).alias("A")
+    assertThrows[SnowparkClientException](
+      df1.join(df2, df1.col("id") === df2.col("id"))
+      .select(df1.col("A.num")))
+  }
```
Comment (sfc-gh-bli) [inline src/test/scala/com/snowflake/snowpark_test/DataFrameAliasSuite.scala:86]: add test for column name contains more than one `.`
For example
```scala
df // contains name `a.b`
val df1 = df.alias("c")
df1.select("c.a.b")
```

**Review by sfc-gh-bli — APPROVED** [review]
Approved with some comments, please merge after address those comments.

**Pull request merged**